src/dsql/Nodes.h

```diff
@@ -194,16 +194,14 @@ class DdlNode : public Node
 
 	// Set the scratch's transaction when executing a node. Fact of accessing the scratch during
 	// execution is a hack.
-	void executeDdl(thread_db* tdbb, DsqlCompilerScratch* dsqlScratch, jrd_tra* transaction)
+	void executeDdl(thread_db* tdbb, DsqlCompilerScratch* dsqlScratch, jrd_tra* transaction, bool trusted = false)
 	{
 		// dsqlScratch should be NULL with CREATE DATABASE.
 		if (dsqlScratch)
 			dsqlScratch->setTransaction(transaction);
 
-		const ULONG flag = checkPermission(tdbb, transaction) ? TDBB_trusted_ddl : 0;
-
-		Firebird::AutoSetRestoreFlag<ULONG> trustedDdlFlag(&tdbb->tdbb_flags, flag, true);
-
+		if (!trusted)
+			checkPermission(tdbb, transaction);
```

Comment:
Shouldn't checkPermission() return void now? And the same question for SCL_check_* routines.

Reply:
Speaking about checkPermission I agree. SCL_check_* require careful investigation. Most of them yes. But some of them return false if object isn't found. It's used while checking CREATE OR ALTER privileges.

Reply:
Then I have a question. If e.g. SCL_check_function() cannot find a function inside RDB$FUNCTIONS, why SCL_check_access() is still called (with non-existing function name and NULL security class)?

Reply:
You're right